app/src/main/java/org/mozilla/focus/biometrics/BiometricAuthenticationHandler.kt

```diff
@@ -0,0 +1,159 @@
+package org.mozilla.focus.biometrics
+
+import android.arch.lifecycle.Lifecycle
+import android.arch.lifecycle.LifecycleObserver
+import android.arch.lifecycle.OnLifecycleEvent
+import android.arch.lifecycle.ProcessLifecycleOwner
+import android.content.Context
+import android.os.Build
+import android.security.keystore.KeyGenParameterSpec
+import android.security.keystore.KeyPermanentlyInvalidatedException
+import android.security.keystore.KeyProperties
+import android.support.annotation.RequiresApi
+import android.support.v4.hardware.fingerprint.FingerprintManagerCompat
+import android.support.v4.os.CancellationSignal
+import java.security.KeyStore
+import java.security.KeyStoreException
+import java.security.NoSuchAlgorithmException
+import java.security.NoSuchProviderException
+import javax.crypto.Cipher
+import javax.crypto.KeyGenerator
+import javax.crypto.SecretKey
+
+@RequiresApi(Build.VERSION_CODES.M)
+class BiometricAuthenticationHandler(
+        private val context: Context): FingerprintManagerCompat.AuthenticationCallback(), LifecycleObserver {
+
+    private var fingerprintManager = FingerprintManagerCompat.from(context)
+    private var cancellationSignal: CancellationSignal? = null
+    private var selfCancelled = false
+    private var keyStore: KeyStore? = null
+    private var keyGenerator: KeyGenerator? = null
+    private var cryptoObject: FingerprintManagerCompat.CryptoObject? = null
+
+    var needsAuth = false; private set
+    var biometricFragment: BiometricAuthenticationDialogFragment? = null; private set
+
+    private val DEFAULT_KEY_NAME = "default_key"
+
+    init {
+        try {
+            keyStore = KeyStore.getInstance("AndroidKeyStore")
+
+        } catch (err: KeyStoreException) {
+            throw RuntimeException("Failed to get instance of KeyStore", err)
+        }
+
+        try {
+            keyGenerator = KeyGenerator.getInstance(KeyProperties.KEY_ALGORITHM_AES, "AndroidKeyStore")
+        } catch (err: NoSuchAlgorithmException) {
+            throw RuntimeException("Failed to get an instance of KeyGenerator", err)
+        } catch (err: NoSuchProviderException) {
+            throw RuntimeException("Failed to get an instance of KeyGenerator", err)
```

Comment:
\[detekt\] ERROR: \[exceptions/ThrowRuntimeException\] Caught exception is too generic\. Prefer catching specific Exceptions to handle specific error cases\.